// ai-interviewer-frontend/src/theme/colors.js
// Theme configuration for the application
export const themeColors = {
  dark: {
    // Background colors
    background: {
      primary: '#282828',
      secondary: '#1a1a1a',
      tertiary: '#404040',
      hover: '#2A2A2A',
    },
    // Text colors
    text: {
      primary: '#cccccc',
      secondary: '#8c8c8c',
      accent: '#ff8c00',
    },
    // Border colors
    border: {
      primary: '#2d2d2d',
      secondary: '#404040',
    },
    // Button colors
    button: {
      primary: {
        background: '#ff8c00',
        hover: '#ff7000',
        text: '#ffffff',
      },
      secondary: {
        background: '#404040',
        hover: '#4a4a4a',
        text: '#cccccc',
      },
      success: {
        background: '#2cbb5d',
        hover: '#2ca854',
        text: '#ffffff',
      },
    },
    // Status colors
    status: {
      success: {
        background: 'rgba(44, 187, 93, 0.1)',
        text: '#2cbb5d',
      },
      error: {
        background: 'rgba(255, 77, 79, 0.1)',
        text: '#ff4d4f',
      },
    },
  },
  light: {
    // Background colors
    background: {
      primary: '#ffffff',
      secondary: '#fafafa',
      tertiary: '#f0f0f0',
      hover: '#f5f5f5',
    },
    // Text colors
    text: {
      primary: '#262626',
      secondary: '#595959',
      accent: '#ff8c00',
    },
    // Border colors
    border: {
      primary: '#e6e6e6',
      secondary: '#d9d9d9',
    },
    // Button colors
    button: {
      primary: {
        background: '#ff8c00',
        hover: '#ff7000',
        text: '#ffffff',
      },
      secondary: {
        background: '#f0f0f0',
        hover: '#e6e6e6',
        text: '#262626',
      },
      success: {
        background: '#2cbb5d',
        hover: '#2ca854',
        text: '#ffffff',
      },
    },
    // Status colors
    status: {
      success: {
        background: 'rgba(44, 187, 93, 0.1)',
        text: '#2cbb5d',
      },
      error: {
        background: 'rgba(255, 77, 79, 0.1)',
        text: '#ff4d4f',
      },
    },
  },
};
